server/zanata-war/src/test/java/org/zanata/adapter/JsonAdapterTest.java

```diff
@@ -23,43 +23,136 @@
     @Test
     public void parseJSON() {
-        Resource resource =
-                adapter.parseDocumentFile(testFile.toURI(), LocaleId.EN,
-                        Optional.absent());
-//        System.out.println(DTOUtil.toXML(resource));
+        Resource resource = parseTestFile("basicjson.json");
         assertThat(resource.getTextFlows()).hasSize(5);
-        assertThat(resource.getTextFlows().get(0).getContents()).isEqualTo(
-                ImmutableList.of("Line One"));
-        // TODO test IDs
+        assertThat(getTextFlowContentsAt(resource, 0)).containsExactly("Line One");
+    }
+
+    /*
+     * JSON parts with duplicated keys are handled separately
+     */
+    @Test
+    public void testStandaloneElementsAreIncluded() {
+        Resource resource = parseTestFile("basicjson.json");
+        assertThat(getTextFlowContentsAt(resource, 3)).containsExactly("First");
+        assertThat(getTextFlowContentsAt(resource, 4)).containsExactly("Second");
+    }
+
+    /*
+     * JSON parts with duplicated keys are separate
+     */
+    @Test
+    public void testDuplicateKeys() {
+        Resource resource = parseTestFile("test-json-duplicateids.json");
+        assertThat(getTextFlowContentsAt(resource, 1)).containsExactly("Line Two");
+        assertThat(getTextFlowContentsAt(resource, 2)).containsExactly("Line Three");
+    }
+
+    /*
+     * JSON parts with duplicated key and content can be considered one entity
+     */
+    @Test
+    public void testDuplicateKeyAndContent() {
+        Resource resource = parseTestFile("test-json-duplicateentries.json");
+        assertThat(resource.getTextFlows()).hasSize(2);
+        assertThat(getTextFlowContentsAt(resource, 0)).containsExactly("Dupe");
+        assertThat(getTextFlowContentsAt(resource, 1)).containsExactly("Same");
+    }
+
+    /*
+     * JSON parts with similar content are separate
+     */
+    @Test
+    public void testDuplicateContent() {
+        Resource resource = parseTestFile("test-json-duplicatecontent.json");
+        assertThat(resource.getTextFlows()).hasSize(3);
+        assertThat(getTextFlowContentsAt(resource, 1)).containsExactly("Same");
+        assertThat(getTextFlowContentsAt(resource, 2)).containsExactly("Same");
+    }
+
+    @Test
+    public void testTranslatedJSONDocument() throws Exception {
+        Map<String, TextFlowTarget> translations = new HashMap<>();
+
+        TextFlowTarget firstTranslation = new TextFlowTarget();
+        firstTranslation.setContents("Foun’dé metalkcta");
+        firstTranslation.setState(ContentState.Approved);
+        translations.put("b76be7dbbd3aaf589a812e1a081d58ca", firstTranslation);
+
+        TextFlowTarget secondTranslation = new TextFlowTarget();
+        secondTranslation.setContents("Tba’dé metalkcta");
+        secondTranslation.setState(ContentState.Translated);
+        translations.put("da4cf86ebff968b8bde6ce056ee2e483", secondTranslation);
```

Comment:
Those are the IDs for the translations, created through the generateHash method.